<html><head><title>Plot Strain Interaction Plots for Exon Array Data</title>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<link rel="stylesheet" type="text/css" href="Rchm.css">
</head>
<body>

<table width="100%"><tr><td>PlotExonResults(ExonModelStrain)</td><td align="right">R Documentation</td></tr></table><object type="application/x-oleobject" classid="clsid:1e2a7bd0-dab9-11d0-b93a-00c04fc99f9e">
<param name="keyword" value="R:   PlotExon">
<param name="keyword" value="R:   PlotExonCI">
<param name="keyword" value="R:   PlotExonResults">
<param name="keyword" value="R:   PlotExonResultsCI">
<param name="keyword" value=" Plot Strain Interaction Plots for Exon Array Data">
</object>


<h2>Plot Strain Interaction Plots for Exon Array Data</h2>


<h3>Description</h3>

<p>
Two functions for plotting Exon Array Data.
</p>


<h3>Usage</h3>

<pre>
  PlotExonResults(resultsframe,ExpSet,dBPackage="mouseexonensembl.db")
  PlotExonResultsCI(resultsframe,ExpSet,dbPackage="mouseexonensembl.db")
  PlotExon(id, ExpSet, display=TRUE, savePlot=FALSE)
  PlotExon(id, ExpSet, display=FALSE, savePlot=TRUE)
  PlotExon(id, ExpSet)
  PlotExonCI(id, ExpSet, display=FALS, savePlot=TRUE)
</pre>


<h3>Arguments</h3>

<table summary="R argblock">
<tr valign="top"><td><code>ExpSet</code></td>
<td>
ExpressionSet containing probeset-level Exon Expression Data.  Currently, only
Mouse Exon 1.0 data is supported.  The ExpressionSet phenoData should contain a
column called 'Strain' where the two strains are coded 1 and 2.
</td></tr>
<tr valign="top"><td><code>id</code></td>
<td>
Valid Ensembl Transcript or Gene ID.  
<br>
Note a list of all valid Transcript or Gene IDs can be queried from the database 
package by using getAllTranscripts() or getAllGenes().
</td></tr>
<tr valign="top"><td><code>resultsframe</code></td>
<td>
Data frame (produced by RunExonModelWorkflow) where there first column contains 
a list of valid Ensembl Transcript or Gene IDs.  
<br>
Note a list of all valid Transcript or Gene IDs can be queried from the database 
package by using getAllTranscripts() or getAllGenes().
</td></tr>
<tr valign="top"><td><code>dbPackage</code></td>
<td>
Name of database package containing Ensembl-Exon Array mapping (currently only 
mouseexonensembl.db exists).
</td></tr>
<tr valign="top"><td><code>savePlot</code></td>
<td>
Boolean. If true, the <code>PlotExon</code> will save the plot in the current working directory.
The name the function will output the name and location for the graphics
file.
</td></tr>
<tr valign="top"><td><code>display</code></td>
<td>
Boolean. If true, the <code>PlotExon</code> will save the plot in the current working directory.
The name the function will output the name and location for the graphics
file.
</td></tr>
</table>

<h3>Details</h3>

<p>
<code>PlotExon</code> produces an interaction plot (in PNG format) for a given Ensembl 
Transcript/Gene ID.  This plot may either be saved to disk in the current working 
directory. <code>PlotExonCI</code> produces a similar plot, but with error bars.  The plots are named automatically using the following convention: 
</p>
<p>
"gene symbol" + "ID" + "exon.plot.png"
</p>
<p>
<code>PlotExonResults</code> produces interaction plots for a data frame of results produced
by <code>RunExonModelWorkflow</code>.  <code>PlotExonResultsCI</code> is
similar, but produces interaction plots with error bars.  Each point in the plot consists of the mean expression of
all probesets for a particular exon for that strain.
</p>
<p>
Note that both <code>PlotExonResults</code> and <code>PlotExon</code> only plots Exons which have 
probesets that exist in the current ExpressionSet, so that Exons that are not represented 
in the data are not shown.  A list of Missing Exons can be found for a particular 
transcript/gene ID by using <code>MissingExons</code>.
</p>


<h3>See Also</h3>

<p>
<code><a href="RunExonModelWorkflow.html">RunExonModelWorkflow</a></code>
<code><a href="MissingExons.html">MissingExons</a></code>
</p>


<h3>Examples</h3>

<pre>
##not run
  #mouseexonensembl.db package required for example
  #load in sample dataset
  data(exontestdata)
  testTrans
  results &lt;- RunExonModelWorkflow(TestSetTrans, testTrans, dBPackage="mouseexonensembl.db")
  #9 out of the 20 transcripts are multiple-exon transcripts
  results$multi
  return
  ##PlotExonResults(results$multi[1:5,], ExpSet=TestSetTrans)
</pre>



<hr><div align="center">[Package <em>ExonModelStrain</em> version 0.2 <a href="00Index.html">Index]</a></div>

</body></html>
